File: web/js/views/freezer.js
// The freezer: a running tab of cooked portions waiting to be eaten (Q24).
// One row per batch, oldest at the top because that is the one to eat next.
// Nothing here touches the plan or the list — freezing is a statement about
// the freezer, and eating from it is not a cooking.

import { api } from "../api.js";
import { confirmDialog, emptyState, fmtRel, foodEmoji, html, openDialog, render, skeleton, toast } from "../dom.js";

// frozen_on is a bare date; give the relative formatter a midday so the
// UTC-stamping in parseUtc can't roll it onto the day before.
const dateRel = (iso) => fmtRel(`${iso}T12:00:00`);
const plural = (n, one, many) => `${n} ${n === 1 ? one : many}`;
const today = () => new Date().toLocaleDateString("en-CA"); // YYYY-MM-DD, local

export async function renderFreezer(root) {
  render(root, skeleton());
  const stock = await api("/freezer");
  const items = stock.items;

  render(root, html`
    <div class="page narrow">
      <div class="page-head">
        <div>
          <h1>The freezer</h1>
          <p class="sub">
            ${items.length === 0
              ? "what's already cooked and waiting"
              : `${plural(stock.total_portions, "portion", "portions")} in ${plural(items.length, "batch", "batches")} · oldest at the top, eat that one first`}
          </p>
        </div>
        <div class="page-actions">
          <button class="btn" data-add>＋ Put something in</button>
        </div>
      </div>

      ${items.length === 0
        ? emptyState(
            "❄️",
            "Nothing in the freezer",
            "When you batch-cook, put the spare portions here and this page becomes the answer to “what's for tea?”.",
          )
        : html`<ul class="row-list">${items.map(batch)}</ul>`}
    </div>
  `);

  root.querySelector("[data-add]").onclick = () => addDialog(root);

  for (const button of root.querySelectorAll("[data-take]")) {
    button.onclick = async () => {
      button.disabled = true;
      try {
        const left = await api(`/freezer/${button.dataset.take}/take`, { method: "POST", body: { portions: 1 } });
        toast(
          left.portions === 0
            ? `That was the last of the ${left.label}.`
            : `One out — ${plural(left.portions, "portion", "portions")} of ${left.label} left.`,
          "ok",
        );
        renderFreezer(root);
      } catch (error) {
        button.disabled = false;
        toast(error.detail || error.message, "error");
      }
    };
  }
  for (const button of root.querySelectorAll("[data-more]")) {
    button.onclick = async () => {
      const item = items.find((i) => i.id === button.dataset.more);
      try {
        await api(`/freezer/${item.id}`, { method: "PATCH", body: { portions: item.portions + 1 } });
        renderFreezer(root);
      } catch (error) {
        toast(error.detail || error.message, "error");
      }
    };
  }
  for (const button of root.querySelectorAll("[data-remove]")) {
    button.onclick = async () => {
      const item = items.find((i) => i.id === button.dataset.remove);
      const ok = await confirmDialog({
        title: `Take out the ${item.label}?`,
        body: `${plural(item.portions, "portion", "portions")} come off the tab — for something binned, given away, or that was never there. Eating one is the −1 button.`,
        confirmLabel: "Take it out",
        danger: true,
      });
      if (!ok) return;
      await api(`/freezer/${item.id}`, { method: "DELETE" });
      toast(`${item.label} is out of the freezer.`, "ok");
      renderFreezer(root);
    };
  }
}

function batch(item) {
  const age = (Date.now() - new Date(`${item.frozen_on}T12:00:00`)) / 86_400_000;
  const name = item.meal_id
    ? html`<a href="#/meals/${item.meal_id}">${item.label}</a>`
    : item.recipe_id
      ? html`<a href="#/recipes/${item.recipe_id}">${item.label}</a>`
      : item.label;
  return html`
    <li class="row-card">
      <span class="rc-emoji" aria-hidden="true">${foodEmoji(item.label)}</span>
      <div class="rc-main">
        <div class="rc-title">${name}</div>
        <div class="rc-sub">
          frozen ${dateRel(item.frozen_on)}${item.note && ` · ${item.note}`}
          ${!item.meal_id && !item.recipe_id && " · not from a recipe here"}
        </div>
      </div>
      <div class="rc-side">
        ${age > 90 && html`<span class="chip butter">been in a while</span>`}
        <span class="chip ${item.portions === 1 ? "red" : "green"}">${plural(item.portions, "portion", "portions")}</span>
        <div class="row-actions">
          <button class="icon-btn" data-take="${item.id}" title="Took one out to eat">−1</button>
          <button class="icon-btn" data-more="${item.id}" title="Recount: there's one more than I said">+1</button>
          <button class="icon-btn warm" data-remove="${item.id}" title="Take the whole batch out">remove</button>
        </div>
      </div>
    </li>
  `;
}

// Three ways to say what went in (Q24): a meal, a recipe, or free text for
// food that never passed through the plan. The picker searches the first two;
// free text is the fallback, not the default.
async function addDialog(root) {
  let kind = "meal";
  let picked = null; // {id, label} from the picker, or null while typing free text
  const dialog = openDialog(html`
    <h2>Put something in the freezer</h2>
    <div class="seg" data-seg>
      <button type="button" class="on" data-kind="meal">A meal</button>
      <button type="button" data-kind="recipe">A recipe</button>
      <button type="button" data-kind="text">Something else</button>
    </div>
    <div data-picker>
      <input type="search" placeholder="Search…" data-q autocomplete="off">
      <ul class="picker-list" data-results></ul>
    </div>
    <form data-f>
      <label class="field" data-text-field hidden><span>What is it</span>
        <input type="text" name="label" maxlength="300" placeholder="Mum's lasagne"></label>
      <div class="form-row">
        <label class="field"><span>Portions</span>
          <input type="number" name="portions" min="1" max="500" step="1" value="4" required></label>
        <label class="field"><span>Frozen on</span>
          <input type="date" name="frozen_on" value="${today()}"></label>
      </div>
      <label class="field"><span>Note <small>(optional)</small></span>
        <input type="text" name="note" maxlength="300" placeholder="the spicy batch"></label>
      <div class="dialog-actions">
        <button class="btn ghost" type="button" data-x>Cancel</button>
        <button class="btn" type="submit" data-submit disabled>Pick one above</button>
      </div>
    </form>
  `);
  dialog.querySelector("[data-x]").onclick = () => dialog.close();

  const picker = dialog.querySelector("[data-picker]");
  const query = dialog.querySelector("[data-q]");
  const results = dialog.querySelector("[data-results]");
  const textField = dialog.querySelector("[data-text-field]");
  const submit = dialog.querySelector("[data-submit]");
  const labelInput = dialog.querySelector('input[name="label"]');

  const refreshSubmit = () => {
    const ready = kind === "text" ? labelInput.value.trim().length > 0 : picked !== null;
    submit.disabled = !ready;
    submit.textContent = ready ? `Freeze ${picked ? picked.label : labelInput.value.trim()}` : kind === "text" ? "Name it first" : "Pick one above";
  };

  const search = async () => {
    const q = query.value.trim() || undefined;
    const rows =
      kind === "meal"
        ? (await api("/meals", { query: { search: q } })).map((m) => ({
            id: m.id,
            label: m.name,
            sub: m.recipes.map((r) => r.title).join(" + ") || "loose ingredients",
          }))
        : (await api("/recipes", { query: { search: q } })).map((r) => ({
            id: r.id,
            label: r.title,
            sub: r.servings ? `serves ${r.servings}` : "",
          }));
    render(results, html`
      ${rows.length === 0 && html`<li><span class="p-sub">Nothing matches — try “Something else” for free text.</span></li>`}
      ${rows.map(
        (row) => html`
          <li>
            <span aria-hidden="true">${foodEmoji(row.label)}</span>
            <div class="p-main">
              <div class="p-title">${row.label}</div>
              ${row.sub && html`<div class="p-sub">${row.sub}</div>`}
            </div>
            ${picked?.id === row.id
              ? html`<span class="chip green">picked</span>`
              : html`<button type="button" class="btn small" data-pick="${row.id}">Pick</button>`}
          </li>
        `,
      )}
    `);
    for (const button of results.querySelectorAll("[data-pick]")) {
      button.onclick = () => {
        picked = rows.find((r) => r.id === button.dataset.pick);
        refreshSubmit();
        search();
      };
    }
  };

  for (const button of dialog.querySelectorAll("[data-kind]")) {
    button.onclick = () => {
      kind = button.dataset.kind;
      picked = null;
      for (const other of dialog.querySelectorAll("[data-kind]")) other.classList.toggle("on", other === button);
      picker.hidden = kind === "text";
      textField.hidden = kind !== "text";
      refreshSubmit();
      if (kind === "text") labelInput.focus();
      else {
        search();
        query.focus();
      }
    };
  }
  labelInput.oninput = refreshSubmit;
  let timer;
  query.oninput = () => {
    clearTimeout(timer);
    timer = setTimeout(search, 220);
  };

  dialog.querySelector("[data-f]").onsubmit = async (event) => {
    event.preventDefault();
    const data = new FormData(event.target);
    const body = { portions: Number(data.get("portions")) || 1 };
    if (data.get("note").trim()) body.note = data.get("note").trim();
    if (data.get("frozen_on")) body.frozen_on = data.get("frozen_on");
    if (kind === "text") body.label = data.get("label").trim();
    else if (kind === "meal") body.meal_id = picked.id;
    else body.recipe_id = picked.id;
    submit.disabled = true;
    try {
      const item = await api("/freezer", { method: "POST", body });
      dialog.close();
      toast(`In it goes — ${plural(item.portions, "portion", "portions")} of ${item.label}.`, "ok");
      renderFreezer(root);
    } catch (error) {
      submit.disabled = false;
      toast(error.detail || error.message, "error");
    }
  };

  await search();
  query.focus();
}

File: backend/app/routers/freezer.py
"""Freezer stock (decision Q24): a running tab of what is in the freezer, in
batches, each linked to the meal or recipe it came from when it came from the
app and free text when it didn't.

Nothing here touches the plan or the shopping list. Freezing a batch is a
statement about the freezer, not about the week, and eating from it is not a
cooking — the cooked record was written when the batch was made.
"""

import uuid

from fastapi import APIRouter, HTTPException, status

from app.deps import CurrentUser, DbSession
from app.routers.meals import get_meal
from app.schemas.freezer import FreezerAddIn, FreezerItemOut, FreezerItemUpdate, FreezerOut, FreezerTakeIn
from app.serializers import freezer_item_out
from app.services.catalog import get_recipe
from app.services.freezer import add_batch, get_item, list_freezer, take_portions

router = APIRouter(prefix="/freezer", tags=["freezer"])

NOT_FOUND = "that batch is not in the freezer; list what is via GET /freezer"


@router.get("", response_model=FreezerOut)
async def get_freezer(user: CurrentUser, db: DbSession) -> FreezerOut:
    """What is in the freezer, oldest batch first — that is the one to eat
    next. Each batch says how many portions are left and, when it came from
    the app, which meal or recipe it was; `total_portions` is the whole
    freezer in one number."""
    items = await list_freezer(db, user.household_id)
    return FreezerOut(items=[freezer_item_out(item) for item in items], total_portions=sum(i.portions for i in items))


@router.post("", response_model=FreezerItemOut, status_code=status.HTTP_201_CREATED)
async def add_to_freezer(payload: FreezerAddIn, user: CurrentUser, db: DbSession) -> FreezerItemOut:
    """Put a batch in the freezer. Name it one way: `meal_id` for a meal
    (the batch takes the meal's name), `recipe_id` for a recipe (the title),
    or `label` for anything that did not come through the app — "mum's
    lasagne", "chicken stock". `portions` is how many are going in
    (default 1); `frozen_on` defaults to today.

    Every call adds a new batch, even for a dish already in there: two
    batches frozen a month apart are two things to eat oldest-first, and
    GET /freezer totals them. To correct a count use PATCH /freezer/{item_id}."""
    if payload.meal_id is not None:
        meal = await get_meal(db, user.household_id, payload.meal_id)
        if meal is None:
            raise HTTPException(status_code=422, detail=f"meal {payload.meal_id} not found; list meals via GET /meals")
        label, meal_id, recipe_id = meal.name, meal.id, None
    elif payload.recipe_id is not None:
        recipe = await get_recipe(db, user.household_id, payload.recipe_id)
        if recipe is None:
            raise HTTPException(
                status_code=422, detail=f"recipe {payload.recipe_id} not found; browse the library via GET /recipes"
            )
        label, meal_id, recipe_id = recipe.title, None, recipe.id
    else:
        assert payload.label is not None  # the schema validator guarantees exactly one
        if not payload.label.strip():
            raise HTTPException(status_code=422, detail="label cannot be blank; say what the batch is")
        label, meal_id, recipe_id = payload.label, None, None
    item = await add_batch(
        db,
        user.household_id,
        label=label,
        portions=payload.portions,
        meal_id=meal_id,
        recipe_id=recipe_id,
        note=payload.note,
        frozen_on=payload.frozen_on,
        user_id=user.id,
    )
    await db.commit()
    return freezer_item_out(item)


@router.patch("/{item_id}", response_model=FreezerItemOut)
async def update_freezer_item(
    item_id: uuid.UUID, payload: FreezerItemUpdate, user: CurrentUser, db: DbSession
) -> FreezerItemOut:
    """Correct a batch: set `portions` outright (a recount), rename it, change
    the note or the date it went in. To eat from it use
    POST /freezer/{item_id}/take instead, which also clears an emptied batch."""
    item = await get_item(db, user.household_id, item_id)
    if item is None:
        raise HTTPException(status_code=404, detail=NOT_FOUND)
    if payload.label is not None:
        if not payload.label.strip():
            raise HTTPException(status_code=422, detail="label cannot be blank; say what the batch is")
        item.label = payload.label.strip()
    if payload.portions is not None:
        item.portions = payload.portions
    if "note" in payload.model_fields_set:
        item.note = (payload.note or "").strip() or None
    if payload.frozen_on is not None:
        item.frozen_on = payload.frozen_on
    await db.commit()
    return freezer_item_out(item)


@router.post("/{item_id}/take", response_model=FreezerItemOut)
async def take_from_freezer(
    item_id: uuid.UUID, payload: FreezerTakeIn, user: CurrentUser, db: DbSession
) -> FreezerItemOut:
    """Take portions out to eat (default 1). The reply is the batch with what
    is left; when that reaches 0 the batch is gone from the freezer and the
    reply is the last you will hear of it. Asking for more than is there
    takes what is there — it is how "we finished the chilli" is said."""
    item = await get_item(db, user.household_id, item_id)
    if item is None:
        raise HTTPException(status_code=404, detail=NOT_FOUND)
    snapshot = freezer_item_out(item)  # taken before the row may be deleted
    remaining = await take_portions(db, item, payload.portions)
    await db.commit()
    return snapshot.model_copy(update={"portions": remaining})


@router.delete("/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_from_freezer(item_id: uuid.UUID, user: CurrentUser, db: DbSession) -> None:
    """Take a whole batch out, whatever is left of it — binned, given away, or
    never there. Nothing else is touched."""
    item = await get_item(db, user.household_id, item_id)
    if item is None:
        raise HTTPException(status_code=404, detail=NOT_FOUND)
    await db.delete(item)
    await db.commit()
